libraries/standard/mqtt/src/mqtt.c

```diff
@@ -301,6 +301,245 @@ static MQTTStatus_t validatePublishParams( const MQTTContext_t * pContext,
+        if( ( nameIndex == ( topicNameLength - 1U ) ) &&
+            ( filterIndex == ( topicFilterLength - 2U ) ) )
+        {
+            /* Determine if the topic filter contains "/+" or "/#" as the last 2 characters.
+             * This covers the special case of topic matching when the topic name
+             * ends with '/' but the topic filter ends with "/+" or "/#".
+             * Thus, for example, topic filter "sport/+" matches the "sport/" but not "sport",
+             * and topic filter "sport/#" matches both "sport/" and "sport". */
+            matchFound = ( ( pTopicFilter[ filterIndex ] == '/' ) &&
+                           ( ( pTopicFilter[ filterIndex + 1U ] == '+' ) ||
+                             ( pTopicFilter[ filterIndex + 1U ] == '#' ) ) ) ? true : false;
+        }
+    }
+
+    return matchFound;
+}
+
+/*-----------------------------------------------------------*/
+
+static bool matchWildcards( const char * pTopicFilter,
+                            uint16_t topicFilterLength,
+                            uint16_t filterIndex,
+                            const char * pTopicName,
+                            uint16_t topicNameLength,
+                            uint16_t * pNameIndex,
+                            bool * pMatch )
+{
+    bool shouldStopMatching = false;
+
+    assert( pTopicFilter != NULL );
+    assert( pTopicName != NULL );
+    assert( pNameIndex != NULL );
+    assert( pMatch != NULL );
+
+    /* Check for '+' wildcard character and verify that it is either at the starting position
+     * or is preceded by a '/'. */
+    if( ( pTopicFilter[ filterIndex ] == '+' ) &&
+        ( ( filterIndex == 0u ) || ( pTopicFilter[ filterIndex - 1U ] == '/' ) ) )
+    {
+        /* Move topic name index to the end of the current level.
+         * This is identified by '/'. */
+        while( ( *pNameIndex < topicNameLength ) && ( pTopicName[ *pNameIndex ] != '/' ) )
+        {
+            ( *pNameIndex )++;
+        }
+
+        /* Decrement the topic name index for 2 different cases:
+         * - If the break condition is ( *pNameIndex < topicNameLength ), then
+         * we have reached the end of the topic name.
+         * - If the break condition is ( pTopicName[ *pNameIndex ] != '/' ),
+         * we move back the index on the '/' character to be at the last
+         * position in the current topic level. */
+        ( *pNameIndex )--;
+    }
+
+    /* Check for '#' wildcard character and that the topic filter has the valid
+    * format for the '#' character.
+    * Note: '#' should be the last character in a topic filter. If the
+    * topic filter has multiple level, "/#" should be the last 2 characters. */
+    else if( ( pTopicFilter[ filterIndex ] == '#' ) &&
+             ( filterIndex == ( topicFilterLength - 1U ) ) && /* '#' should be the last character. */
+             ( ( filterIndex == 0u ) || ( pTopicFilter[ filterIndex - 1U ] == '/' ) ) )
```

Comment:
This looks kind of excessive. Some intermediate boolean variables might be more helpful than comments here in keeping everything readable. If we're validating the filter the same time as we're matching, would a `isValidFilter()` function be helpful to call before we start matching?

Some suggestions to improve hygiene, don't necessarily have to be in this PR. I'm going to approve the PR with the assumption you've run the coverage tests and verified all these lines are covered